src/util/s3select.js

```diff
@@ -0,0 +1,115 @@
+/* Copyright (C) 2023 NooBaa */
+'use strict';
+
+const dbg = require('./debug_module')(__filename);
+const { Transform } = require('stream');
+const nb_native = require('./nb_native');
+const assert = require('assert');
+
+/*Encodes an s3select response according to AWS format. See
+https://docs.aws.amazon.com/AmazonS3/latest/API/RESTSelectObjectAppendix.html*/
+
+class S3SelectStream extends Transform {
+
+    static prelude_length_bytes = 12;
+    static crc_length_bytes = 4;
+    static string_type = 7;
+    static header_max_len = 256;
+
+    static header_to_buffer(key, val) {
+        assert(key.length < S3SelectStream.header_max_len);
+        assert(val.length < S3SelectStream.header_max_len);
+        return Buffer.concat([
+            Buffer.from([key.length]),
+            Buffer.from(key),
+            Buffer.from([S3SelectStream.string_type, 0, val.length]),
+            Buffer.from(val)
+        ]);
+    }
+
+    //end message is always the same, so we allocate and init it once.
+    static end_message = Buffer.concat([
+        Buffer.from([0, 0, 0, 56]), //PRELUDE begins: total byte length
+        Buffer.from([0, 0, 0, 40]), //headers byte length
+        Buffer.from([0xc1, 0xc6, 0x84, 0xd4]), //crc of above (end of PRELUDE)
+        //DATA begins, HEADERS begin
+        S3SelectStream.header_to_buffer(':event-type', 'End'),
+        S3SelectStream.header_to_buffer(':message-type', 'event'),
+        //no PAYLOAD (DATA ends)
+        Buffer.from([0xfe, 0x2c, 0xee, 0x99]) //MESSAGE CRC (crc of above)
+    ]);
+
+    //same for headers of records message
+    static records_message_headers = Buffer.concat([
+        S3SelectStream.header_to_buffer(':event-type', 'Records'),
+        S3SelectStream.header_to_buffer(':content-type', 'application/octet-stream'),
+        S3SelectStream.header_to_buffer(':message-type', 'event'),
+    ]);
+
+    constructor(opts) {
```

Comment:
Can you defined in a jsdoc what opts can be provided:
```suggestion
    /**
     * @param {{
     *     query: string;
     *     input_format: 'CSV' | 'JSON';
     *     input_serialization_format: {
     *        FieldDelimiter: string;
     *        RecordDelimiter: string;
     *        FileHeaderInfo: string;
     *     };
     *     records_header_buf: Buffer;
     *     ...
     *  }} opts
     */
    constructor(opts) {
```